src/Array.mo

```diff
@@ -119,25 +118,70 @@ module {
   };
   /// Output array contains each array-value if and only if the predicate is true; ordering retained.
   public func filter<A>(xs : [A], f : A -> Bool) : [A] {
-    let ys : Buffer.Buffer<A> = Buffer.Buffer(xs.size());
-    for (x in xs.vals()) {
-      if (f(x)) {
-        ys.add(x);
-      };
-    };
-    ys.toArray();
+    var count = 0;
```

Comment:
What's the rationale for the lower-level implementation? Just two 2 array allocations versus repeated grows? Is that always a win? I guess its much better with our GC at then end implementation.

Reply:
It was to remove the dependency of the Array module on the Buffer module (was becoming circular, and I reasoned that since Buffers abstract over Arrays, the dependency should be removed from the Array module).

I just wrote it as performant as I could as a secondary thought, and to not degrade performance from before. Although I believe this implementation is more efficient because the old implementation had all the resizes (which strictly subsumes one of the array allocations in the new implementation), and the final conversion to array (which is equivalent the the second allocation in the new implementation).

Reply:
Makes sense.